### PR TITLE
Support Cron Schedule as part of the TestWorkflowEnvironment

### PR DESCRIPTION
**Pull request by assafsch**
**What changed?**
Allows setting a Cron schedule for non-child Workflows in the Test environment using
`SetWorkflowCronSchedule` and `SetWorkflowCronMaxIterations`.

**Why?**
I wanted to be able to run Cron Workflows and not have to wrap them inside the test Workflow.

**How did you test it?**
Yes, local testing via unit tests + made sure my code can use the functionality

**Potential risks**
Nothing, this is related only to the Workflow testing environment

### CODE REVIEW

**Comment by yux0** [comment]
Any reason to support the cron in this test suite?

**Comment by assafsch** [comment]
We have a use-case where we use a Cron Workflow that relies on previous iterations.
With the current test-suite, we cannot test how our workflow behaves, since each call to `ExecuteWorkflow` resets everything and the Workflow is considered a clean slate.

Plus, since it's a functionality that the Go SDK offers, why not allow for such complementing functionality in the test-suite? Am I missing something?

File: internal/internal_workflow_testsuite.go [hunk 1]
```diff
@@ -183,6 +185,11 @@ type (
 
 		workerStopChannel  chan struct{}
 		sessionEnvironment *testSessionEnvironmentImpl
+
+		cronSchedule       string
```
Comment (yux0) [inline internal/internal_workflow_testsuite.go:189]: Store the cron schedule in env.workflowInfo.CronSchedule should be enough?

File: internal/internal_workflow_testsuite.go [hunk 2]
```diff
@@ -246,13 +255,19 @@ func newTestWorkflowEnvironmentImpl(s *WorkflowTestSuite, parentRegistry *regist
...
+		cronIterations: 0,
 	}
 
 	// move forward the mock clock to start time.
 	env.setStartTime(time.Now())
 
 	// put current workflow as a running workflow so child can send signal to parent
-	env.runningWorkflows[env.workflowInfo.WorkflowExecution.ID] = &testWorkflowHandle{env: env, callback: func(result []byte, err error) {}}
+	testWorkflowHandle := &testWorkflowHandle{env: env, callback: func(result []byte, err error) {}}
+	if len(env.cronSchedule) > 0 {
```
[the remark this answers is not in the document]
Comment (assafsch) [reply internal/internal_workflow_testsuite.go:267]: depending on the actual Test use-case.
A user would have to call `SetWorkflowCronSchedule`, which would set the cron string to be > 0.

File: internal/internal_workflow_testsuite.go [hunk 3]
```diff
@@ -916,7 +998,7 @@ func (env *testWorkflowEnvironmentImpl) GetLogger() *zap.Logger {
 	return env.logger
 }
 
-func (env *testWorkflowEnvironmentImpl) GetMetricsScope() tally.Scope {
+ func (env *testWorkflowEnvironmentImpl) GetMetricsScope() tally.Scope {
```
[the remark this answers is not in the document]
Comment (assafsch) [reply internal/internal_workflow_testsuite.go:1001]: yep, I should :)

File: internal/internal_workflow_testsuite.go [hunk 4]
```diff
@@ -800,7 +828,12 @@ func (env *testWorkflowEnvironmentImpl) Complete(result []byte, err error) {
 		env.testResult = newEncodedValue(result, dc)
 	}
 
-	close(env.doneChannel)
+	// Only close on:
+	// 1. Child-Workflows
+	// 2. non-cron Workflows
+	if !env.isChildWorkflow() && !env.IsCron() {
```
[the remark this answers is not in the document]
Comment (assafsch) [reply internal/internal_workflow_testsuite.go:834]: hmm... let me check

**Comment by assafsch** [comment]
Thanks @yux0 for looking into this. I believe all of the comments should be addressed now.

**Review by yux0 — APPROVED** [review]

**Comment by assafsch** [comment]
@yux0 seems like there are merge conflicts for the IDLs. I may be out of my depth here :)

**Pull request merged**